File: findproxylists.clj
#!/usr/bin/env boot

(set-env! :dependencies '[[com.taoensso/timbre "4.10.0"]])

(require '[boot.cli :refer [defclifn]])
(require '[com.taoensso/timber :refer [spy]])

(def PROXY-LIST_DIR "wsglists")
(def NEEDED-CARDS-FILE "needed-cards")

(defn read-wsg-lists []
  (let [dir    (clojure.java.io/file PROXY-LIST_DIR)
        files  (map #(str %)
                    (.list dir))
        filenames  (map #(str PROXY-LIST_DIR "/" %) 
                        (.list dir))
        ;;_ (clojure.pprint/pprint (type files))
        ;;_ (clojure.pprint/pprint (type filenames))
        ]
    (into {}
          (map #(with-open [rdr  (clojure.java.io/reader (str PROXY-LIST_DIR "/" %))]
                  {% (into [] (line-seq rdr))})
               files))
    ))

(defn read-req-cards []
  (let [cards   (->> (with-open [rdr (clojure.java.io/reader NEEDED-CARDS-FILE)]
                       (into [] (line-seq rdr)))
                     (filter #(not (= "" %))))
        ;;_ (clojure.pprint/pprint cards)
        ]
   cards
   ))


(defn find-lists-with-card
  [lists-map card]
(let [;;_ (clojure.pprint/pprint "lists-map:")
      ;;_ (clojure.pprint/pprint lists-map)
      _ (clojure.pprint/pprint "card:")
      _ (clojure.pprint/pprint card)
      _ (clojure.pprint/pprint (type lists-map))
      _ (clojure.pprint/pprint (keys lists-map))
      

      filtered (filter (comp (timbre/spy #(.contains % card)) last) lists-map)
      filteredtest (.contains (get lists-map "wsglists/Master") card)
      _ (clojure.pprint/pprint "Filtered:")
      _ (clojure.pprint/pprint filteredtest)
      ;;_ (clojure.pprint/pprint filtered)
      filtered-names (map first filtered)
      _ (clojure.pprint/pprint filtered-names)
      ]
  (filter (comp #(.contains % card) last) lists-map)  
  ))





(defn -main [& args]
  (let [wsg-list-map (read-wsg-lists)
        ;;_ (clojure.pprint/pprint wsg-list-map)
        req-cards    (read-req-cards)
        a-card       (first req-cards)

        _ (clojure.pprint/pprint a-card)
        _ (clojure.pprint/pprint (find-lists-with-card wsg-list-map a-card))
        ;;list-with-card (filter #())
        
        ;;_ (clojure.pprint/pprint wsg-list-map)
        ;;_ (clojure.pprint/pprint (vals wsg-list-map))
        ]
    #_(clojure.pprint/pprint (find-lists-with-card wsg-list-map a-card))
    )
)

File: findproxylists.rb
#!/usr/bin/env ruby

require 'pp'

include_master = false

def read_wsg_lists()
    lists = {}
    Dir.foreach('wsglists') do |file_list|
        next if file_list == '.' or file_list == '..'
        list_cards = File.readlines("wsglists/" + file_list)
        lists[file_list] = list_cards.each {|line| line.strip!}
    end
    lists
end

def read_req_cards()
    needed_cards = []
    File.readlines("needed-cards").each do |line|
        line.strip!
        next if line == '' or line.include? "#"
        needed_cards << line
    end
    needed_cards
end

def find_list_with_card(list_map, card)
    matching_lists = []
    list_map.each {|list_name, card_list|
        matching_lists << list_name if card_list.include? card
    }
    matching_lists
end

def find_lists_foreach_card(list_map, cards) 
    card_list_map = {}
    cards.each {|card|
        lists = find_list_with_card(list_map, card)
        card_list_map[card] = lists
    }
    card_list_map
end

def print_list(cards_list_map)
    cards_list_map.each {|key, value|
        print '%-30s' % key
        value.each {|list|
            print '%-25s' % list
        }
        puts ""
    }
   puts  
end


req_cards = read_req_cards
wsg_lists = read_wsg_lists
wsg_lists.delete("Master") if !include_master

card_lists = find_lists_foreach_card(wsg_lists, req_cards)
print_list(card_lists)